// @fuel-ts/abi-typegen/dist/cli.js
"use strict";var ne=Object.create;var j=Object.defineProperty;var ie=Object.getOwnPropertyDescriptor;var se=Object.getOwnPropertyNames;var oe=Object.getPrototypeOf,pe=Object.prototype.hasOwnProperty;var ae=(n,t,e)=>t in n?j(n,t,{enumerable:!0,configurable:!0,writable:!0,value:e}):n[t]=e;var ce=(n,t)=>{for(var e in t)j(n,e,{get:t[e],enumerable:!0})},dt=(n,t,e,r)=>{if(t&&typeof t=="object"||typeof t=="function")for(let i of se(t))!pe.call(n,i)&&i!==e&&j(n,i,{get:()=>t[i],enumerable:!(r=ie(t,i))||r.enumerable});return n};var J=(n,t,e)=>(e=n!=null?ne(oe(n)):{},dt(t||!n||!n.__esModule?j(e,"default",{value:n,enumerable:!0}):e,n)),me=n=>dt(j({},"__esModule",{value:!0}),n);var p=(n,t,e)=>(ae(n,typeof t!="symbol"?t+"":t,e),e);var ge={};ce(ge,{configureCliOptions:()=>ee,resolveProgramType:()=>Zt,run:()=>he,runCliAction:()=>te});module.exports=me(ge);var Wt=require("@fuel-ts/versions"),k=require("commander");var jt=require("@ethersproject/bytes"),_=require("fs"),Jt=require("glob"),qt=J(require("mkdirp")),Kt=require("path"),Yt=J(require("rimraf"));var It=J(require("lodash.upperfirst"));function ht(n){let e=[r=>r.replace(/\s+/g,"-"),r=>r.replace(/\./g,"-"),r=>r.replace(/_/g,"-"),r=>r.replace(/-[a-z]/g,i=>i.slice(-1).toUpperCase()),r=>r.replace(/-/g,""),r=>r.replace(/^\d+/,""),r=>(0,It.default)(r)].reduce((r,i)=>i(r),n);if(e==="")throw new Error(`Can't guess class name, please rename file: ${n}`);return e}function I(n){let{types:t,typeId:e}=n,r=t.find(({rawAbiType:{typeId:i}})=>i===e);if(!r)throw new Error(`Type ID not found: ${e}.`);return r.parseComponentsAttributes({types:t}),r}function h(n){let{types:t,typeArguments:e,parentTypeId:r,target:i}=n,o=`${i}Label`,s=[],a,m;r!==void 0&&(a=I({types:t,typeId:r}),m=a.attributes[o]),e.forEach(y=>{let b,d=y.type;try{b=I({types:t,typeId:d}).attributes[o]}catch{b="void"}if(y.typeArguments){let l=h({types:t,target:i,parentTypeId:y.type,typeArguments:y.typeArguments});s.push(l)}else s.push(`${b}`)});let u=s.join(", ");return m&&(u=`${m}<${u}>`),u}var Y=class{name;types;rawAbiFunction;attributes;constructor(t){this.rawAbiFunction=t.rawAbiFunction,this.types=t.types,this.name=t.rawAbiFunction.name,this.attributes={inputs:this.bundleInputTypes(),output:this.bundleOutputTypes(),prefixedInputs:this.bundleInputTypes(!0)}}bundleInputTypes(t=!1){let{types:e}=this;return this.rawAbiFunction.inputs.map(i=>{let{name:o,type:s,typeArguments:a}=i,m=I({types:e,typeId:s}),u;return a?u=h({types:e,target:"input",parentTypeId:s,typeArguments:a}):u=m.attributes.inputLabel,t?`${o}: ${u}`:u}).join(", ")}bundleOutputTypes(){return h({types:this.types,target:"output",typeArguments:[this.rawAbiFunction.output]})}getDeclaration(){let{name:t}=this,{prefixedInputs:e,output:r}=this.attributes;return`${t}: InvokeFunction<[${e}], ${r}>`}};function gt(n){let{types:t,rawAbiFunction:e}=n;return new Y({types:t,rawAbiFunction:e})}function At(n){let{types:t,rawAbiFunctions:e}=n;return e.map(i=>gt({types:t,rawAbiFunction:i}))}var T=class{rawAbiType;attributes;requireImportFromFuels;constructor(t){this.rawAbiType=t.rawAbiType,this.attributes={inputLabel:"unknown",outputLabel:"unknown"},this.requireImportFromFuels=!1}};var W=class extends T{name="array";static isSuitableFor(t){return W.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){var u,y;let{types:e}=t,{type:r}=this.rawAbiType,i=Number((u=r.match(W.MATCH_REGEX))==null?void 0:u[1]),o=[],s=[];(y=this.rawAbiType.components)==null||y.forEach(b=>{let{type:d,typeArguments:l}=b;if(l){let f=h({types:e,typeArguments:l,parentTypeId:d,target:"input"}),x=h({types:e,typeArguments:l,parentTypeId:d,target:"output"});o.push(f),s.push(x)}else{let{attributes:f}=I({types:e,typeId:d});o.push(f.inputLabel),s.push(f.outputLabel)}});let a=Array(i).fill(o[0]).join(", "),m=Array(i).fill(s[0]).join(", ");return this.attributes={inputLabel:`[${a}]`,outputLabel:`[${m}]`},this.attributes}},E=W;p(E,"swayType","[_; 2]"),p(E,"MATCH_REGEX",/^\[_; ([0-9]+)\]$/m);var st=class extends T{name="str";static isSuitableFor(t){return st.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){return this.attributes={inputLabel:"string",outputLabel:"string"},this.attributes}},w=st;p(w,"swayType","str[3]"),p(w,"MATCH_REGEX",/^str\[(.+)\]$/m);var ot=class extends w{name="b256";static isSuitableFor(t){return ot.MATCH_REGEX.test(t.type)}},F=ot;p(F,"swayType","b256"),p(F,"MATCH_REGEX",/^b256$/m);var pt=class extends F{name="b512";static isSuitableFor(t){return pt.MATCH_REGEX.test(t.type)}},$=pt;p($,"swayType","struct B512"),p($,"MATCH_REGEX",/^struct B512$/m);var at=class extends T{name="bool";static isSuitableFor(t){return at.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){return this.attributes={inputLabel:"boolean",outputLabel:"boolean"},this.attributes}},O=at;p(O,"swayType","bool"),p(O,"MATCH_REGEX",/^bool$/m);function L(n){var i;let{rawAbiType:t,regex:e}=n,r=(i=t.type.match(n.regex))==null?void 0:i[1];if(!r){let o=`Couldn't extract struct name with: '${e}'.

`;throw o+=`Check your JSON ABI.

[source]
`,o+=`${JSON.stringify(t,null,2)}`,new Error(o)}return r}var q=class extends T{name="enum";static isSuitableFor(t){let e=q.MATCH_REGEX.test(t.type),r=q.IGNORE_REGEX.test(t.type);return e&&!r}parseComponentsAttributes(t){let e=this.getStructName();return this.attributes={structName:e,inputLabel:`${e}Input`,outputLabel:`${e}Output`},this.attributes}getStructName(){return L({rawAbiType:this.rawAbiType,regex:q.MATCH_REGEX})}getStructContents(t){let{types:e,target:r}=t,{components:i}=this.rawAbiType,o=i,s=`${r}Label`;return o.map(m=>{let{name:u,type:y}=m;if(y===0)return`${u}: []`;let{attributes:b}=I({types:e,typeId:y});return`${u}: ${b[s]}`}).join(", ")}},C=q;p(C,"swayType","enum MyEnumName"),p(C,"MATCH_REGEX",/^enum (.+)$/m),p(C,"IGNORE_REGEX",/^enum Option$/m);var Z=class extends T{name="generic";static isSuitableFor(t){return Z.MATCH_REGEX.test(t.type)}getStructName(){return L({rawAbiType:this.rawAbiType,regex:Z.MATCH_REGEX})}parseComponentsAttributes(t){let e=this.getStructName();return this.attributes={inputLabel:e,outputLabel:e},this.attributes}},G=Z;p(G,"swayType","generic T"),p(G,"MATCH_REGEX",/^generic ([^\s]+)$/m);var ct=class extends T{name="option";static isSuitableFor(t){return ct.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){return this.attributes={inputLabel:"Option",outputLabel:"Option"},this.attributes}},P=ct;p(P,"swayType","enum Option"),p(P,"MATCH_REGEX",/^enum Option$/m);var mt=class extends T{name="u8";constructor(t){super(t),this.attributes={inputLabel:"BigNumberish",outputLabel:"number"},this.requireImportFromFuels=!0}static isSuitableFor(t){return mt.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){return this.attributes}},g=mt;p(g,"swayType","u8"),p(g,"MATCH_REGEX",/^u8$/m);var ut=class extends g{name="u64";parseComponentsAttributes(t){return this.attributes={inputLabel:"BigNumberish",outputLabel:"BN"},this.attributes}static isSuitableFor(t){return ut.MATCH_REGEX.test(t.type)}},R=ut;p(R,"swayType","u64"),p(R,"MATCH_REGEX",/^u64$/m);var yt=class extends R{name="rawUntypedPtr";static isSuitableFor(t){return yt.MATCH_REGEX.test(t.type)}},M=yt;p(M,"swayType","raw untyped ptr"),p(M,"MATCH_REGEX",/^raw untyped ptr$/m);var K=class extends T{name="struct";static isSuitableFor(t){let e=K.MATCH_REGEX.test(t.type),r=K.IGNORE_REGEX.test(t.type);return e&&!r}parseComponentsAttributes(t){let e=this.getStructName();return this.attributes={structName:e,inputLabel:`${e}Input`,outputLabel:`${e}Output`},this.attributes}getStructName(){return L({rawAbiType:this.rawAbiType,regex:K.MATCH_REGEX})}getStructContents(t){let{types:e,target:r}=t,{components:i}=this.rawAbiType;return i.map(a=>{let{name:m,type:u,typeArguments:y}=a,b=I({types:e,typeId:u}),d;if(y)d=h({types:e,target:r,parentTypeId:u,typeArguments:y});else{let l=`${r}Label`;d=b.attributes[l]}return`${m}: ${d}`}).join(", ")}getStructDeclaration(t){let{types:e}=t,{typeParameters:r}=this.rawAbiType;return r?`<${r.map(s=>I({types:e,typeId:s})).map(({attributes:{inputLabel:s}})=>s).join(", ")}>`:""}},N=K;p(N,"swayType","struct MyStruct"),p(N,"MATCH_REGEX",/^struct (.+)$/m),p(N,"IGNORE_REGEX",/^struct (Vec|RawVec)$/m);var lt=class extends T{name="tupple";static isSuitableFor(t){return lt.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){var o;let{types:e}=t,r=[],i=[];return(o=this.rawAbiType.components)==null||o.forEach(s=>{let{type:a,typeArguments:m}=s;if(m){let u=h({types:e,typeArguments:m,parentTypeId:a,target:"input"}),y=h({types:e,typeArguments:m,parentTypeId:a,target:"output"});r.push(u),i.push(y)}else{let{attributes:u}=I({types:e,typeId:a});r.push(u.inputLabel),i.push(u.outputLabel)}}),this.attributes={inputLabel:`[${r.join(", ")}]`,outputLabel:`[${i.join(", ")}]`},this.attributes}},H=lt;p(H,"swayType","(_, _, _)"),p(H,"MATCH_REGEX",/^\([_,\s]+\)$/m);var ft=class extends g{name="u16";static isSuitableFor(t){return ft.MATCH_REGEX.test(t.type)}},X=ft;p(X,"swayType","u16"),p(X,"MATCH_REGEX",/^u16$/m);var bt=class extends g{name="u32";static isSuitableFor(t){return bt.MATCH_REGEX.test(t.type)}},U=bt;p(U,"swayType","u32"),p(U,"MATCH_REGEX",/^u32$/m);var tt=class extends E{name="vector";static isSuitableFor(t){let e=tt.MATCH_REGEX.test(t.type),r=tt.IGNORE_REGEX.test(t.type);return e&&!r}parseComponentsAttributes(t){return this.attributes={inputLabel:"Vec",outputLabel:"Vec"},this.attributes}},S=tt;p(S,"swayType","struct Vec"),p(S,"MATCH_REGEX",/^struct Vec/m),p(S,"IGNORE_REGEX",/^struct RawVec$/m);var xt=[E,F,$,O,C,G,P,M,w,N,H,X,U,R,g,S];function Et(n){let{rawAbiType:t}=n,{type:e}=t,r=xt.find(i=>i.isSuitableFor({type:e}));if(!r)throw new Error(`Type not supported: ${e}`);return new r(n)}function wt(n){return["()","struct RawVec"].indexOf(n.type)>=0}function Ft(n){let t=[];return n.rawAbiTypes.forEach(e=>{let{type:r}=e;if(!wt({type:r})){let o=Et({rawAbiType:e});t.push(o)}}),t.forEach(e=>{e.parseComponentsAttributes({types:t})}),t}var et=class{name;programType;filepath;outputDir;commonTypesInUse=[];rawContents;hexlifiedBinContents;types;functions;constructor(t){let{filepath:e,outputDir:r,rawContents:i,hexlifiedBinContents:o,programType:s}=t,a=/([^/]+)-abi\.json$/m,m=e.match(a);if(!m||m.length===0)throw new Error(`Could not parse name from abi file: ${e}`);let y=`${ht(m[1])}Abi`;this.name=y,this.programType=s,this.filepath=e,this.rawContents=i,this.hexlifiedBinContents=o,this.outputDir=r;let{types:b,functions:d}=this.parse();this.types=b,this.functions=d,this.computeCommonTypesInUse()}parse(){let{types:t,functions:e}=this.rawContents,r=Ft({rawAbiTypes:t}),i=At({rawAbiFunctions:e,types:r});return{types:r,functions:i}}computeCommonTypesInUse(){let t={option:"Option",enum:"Enum",vector:"Vec"};this.commonTypesInUse=[],Object.keys(t).forEach(e=>{if(!!this.types.find(i=>i.name===e)){let i=t[e];this.commonTypesInUse.push(i)}})}};var Pt=require("path");var Ct=require("@fuel-ts/versions"),Tt=require("handlebars");var Rt=`/* Autogenerated file. Do not edit manually. */

/* tslint:disable */
/* eslint-disable */

/*
  Fuels version: {{FUELS}}
  Forc version: {{FORC}}
  Fuel-Core version: {{FUEL_CORE}}
*/
`;function A(n){let{data:t,template:e}=n,r={strict:!0,noEscape:!0},i=(0,Tt.compile)(e,r),o=(0,Tt.compile)(Rt,r);return i({...t,header:o(Ct.versions)}).replace(/[\n]{3,}/gm,`

`)}var Nt=`{{header}}

/*
  Mimics Sway Enum, requires at least one Key-Value but
  does not raise error on multiple pairs.
  This is done in the abi-coder
*/
export type Enum<T, U = { [K in keyof T]: Pick<T, K> }> = Partial<T> &
  U[keyof U];

/*
  Mimics Sway Option and Vectors.
  Vectors are treated like arrays in Typescript.
*/
export type Option<T> = T | undefined;

export type Vec<T> = T[];
`;function D(){return A({template:Nt})}var St=`{{header}}

{{#if isGeneratingContracts}}
{{#each abis}}
export type { {{name}} } from './{{name}}';
{{/each}}
{{/if}}

{{#each abis}}
export { {{name}}__factory } from './factories/{{name}}__factory';
{{/each}}
`;function v(n){let{abis:t}=n,e=t[0].programType==="contract";return A({template:St,data:{abis:t,isGeneratingContracts:e}})}function V(n){let{types:t}=n;return{enums:t.filter(r=>r.name==="enum").map(r=>{let i=r,o=i.getStructName(),s=i.getStructContents({types:t,target:"input"}),a=i.getStructContents({types:t,target:"output"});return{structName:o,inputValues:s,outputValues:a,recycleRef:s===a}})}}var _t=J(require("lodash.uniq")),fe=(n,t)=>n.toLowerCase().localeCompare(t.toLowerCase());function z(n){let{types:t,baseMembers:e=[]}=n,i=t.filter(s=>s.requireImportFromFuels).map(s=>s.attributes.inputLabel),o=(0,_t.default)(e.concat(i).sort(fe));return{imports:o.length?o:void 0}}function B(n){let{types:t}=n;return{structs:t.filter(r=>r.name==="struct").map(r=>{let i=r,o=i.getStructName(),s=i.getStructContents({types:t,target:"input"}),a=i.getStructContents({types:t,target:"output"}),m=i.getStructDeclaration({types:t});return{structName:o,typeAnnotations:m,inputValues:s,outputValues:a,recycleRef:s===a}})}}var $t=`{{header}}

{{#if imports}}
import type {
{{#each imports}}
  {{this}},
{{/each}}
} from 'fuels';
{{/if}}

{{#if commonTypesInUse}}
import type { {{commonTypesInUse}} } from "./common";
{{/if}}

{{#each enums}}
export type {{structName}}Input = Enum<{ {{inputValues}} }>;
{{#if recycleRef}}
export type {{structName}}Output = {{structName}}Input;
{{else}}
export type {{structName}}Output = Enum<{ {{outputValues}} }>;
{{/if}}
{{/each}}


{{#each structs}}
export type {{structName}}Input{{typeAnnotations}} = { {{inputValues}} };
{{#if recycleRef}}
export type {{structName}}Output{{typeAnnotations}} = {{structName}}Input{{typeAnnotations}};
{{else}}
export type {{structName}}Output{{typeAnnotations}} = { {{outputValues}} };
{{/if}}
{{/each}}


interface {{capitalizedName}}Interface extends Interface {
  functions: {
    {{#each functionsFragments}}
    {{this}}: FunctionFragment;
    {{/each}}
  };

  {{#each encoders}}
  encodeFunctionData(functionFragment: '{{functionName}}', values: [{{input}}]): Uint8Array;
  {{/each}}

  {{#each decoders}}
  decodeFunctionData(functionFragment: '{{functionName}}', data: BytesLike): DecodedValue;
  {{/each}}
}


export class {{capitalizedName}} extends Contract {
  interface: {{capitalizedName}}Interface;
  functions: {
    {{#each functionsTypedefs}}
    {{this}};
    {{/each}}
  };
}
`;function Ot(n){let{name:t,types:e,functions:r,commonTypesInUse:i}=n.abi,o=r.map(l=>l.getDeclaration()),s=r.map(l=>l.name),a=r.map(l=>({functionName:l.name,input:l.attributes.inputs})),m=r.map(l=>({functionName:l.name})),{enums:u}=V({types:e}),{structs:y}=B({types:e}),{imports:b}=z({types:e,baseMembers:["Interface","FunctionFragment","DecodedValue","Contract","BytesLike","InvokeFunction","BN"]});return A({template:$t,data:{capitalizedName:t,commonTypesInUse:i.join(", "),functionsTypedefs:o,functionsFragments:s,encoders:a,decoders:m,structs:y,enums:u,imports:b}})}var Lt=`{{header}}

import { Interface, Contract } from "fuels";
import type { Provider, Account, AbstractAddress } from "fuels";
import type { {{capitalizedName}}, {{capitalizedName}}Interface } from "../{{capitalizedName}}";

const _abi = {{abiJsonString}}

export class {{capitalizedName}}__factory {
  static readonly abi = _abi
  static createInterface(): {{capitalizedName}}Interface {
    return new Interface(_abi) as unknown as {{capitalizedName}}Interface
  }
  static connect(
    id: string | AbstractAddress,
    accountOrProvider: Account | Provider
  ): {{capitalizedName}} {
    return new Contract(id, _abi, accountOrProvider) as unknown as {{capitalizedName}}
  }
}
`;function Gt(n){let{name:t,rawContents:e}=n.abi,r=JSON.stringify(e,null,2);return A({template:Lt,data:{capitalizedName:t,abiJsonString:r}})}function Mt(n){let{abis:t,outputDir:e}=n,r=[],i=t.find(s=>s.commonTypesInUse.length>0);t.forEach(s=>{let{name:a}=s,m=`${e}/${a}.d.ts`,u=`${e}/factories/${a}__factory.ts`,y={path:m,contents:Ot({abi:s})},b={path:u,contents:Gt({abi:s})};r.push(y),r.push(b)});let o={path:`${e}/index.ts`,contents:v({abis:t})};if(r.push(o),i){let a={path:(0,Pt.join)(e,"common.d.ts"),contents:D()};r.push(a)}return r}var Ut=require("path");var Ht=`{{header}}

{{#if imports}}
import {
{{#each imports}}
  {{this}},
{{/each}}
} from 'fuels';
{{/if}}

{{#if commonTypesInUse}}
import type { {{commonTypesInUse}} } from "./common";
{{/if}}


{{#each enums}}
export type {{structName}}Input = Enum<{ {{inputValues}} }>;
{{#if recycleRef}}
export type {{structName}}Output = {{structName}}Input;
{{else}}
export type {{structName}}Output = Enum<{ {{outputValues}} }>;
{{/if}}
{{/each}}


{{#each structs}}
export type {{structName}}Input{{typeAnnotations}} = { {{inputValues}} };
{{#if recycleRef}}
export type {{structName}}Output{{typeAnnotations}} = {{structName}}Input{{typeAnnotations}};
{{else}}
export type {{structName}}Output{{typeAnnotations}} = { {{outputValues}} };
{{/if}}
{{/each}}

type {{capitalizedName}}Inputs = [{{inputs}}];

const _abi = {{abiJsonString}}

const _bin = '{{hexlifiedBinString}}'

export class {{capitalizedName}}__factory {

  static readonly abi = _abi
  static readonly bin = _bin;

  static createInstance(provider?: Provider) {

    const { abi, bin } = {{capitalizedName}}__factory

    const predicate = new Predicate(bin, abi, provider);

    return predicate;

  }

}
`;function Xt(n){let{abi:t}=n,{types:e}=t,{rawContents:r,name:i,hexlifiedBinContents:o}=n.abi,s=JSON.stringify(r,null,2),a=t.functions.find(f=>f.name==="main");if(!a)throw new Error("ABI doesn't have a 'main()' method.");let{enums:m}=V({types:e}),{structs:u}=B({types:e}),{imports:y}=z({types:e,baseMembers:["Predicate","Provider"]}),{prefixedInputs:b,output:d}=a.attributes;return A({template:Ht,data:{inputs:b,output:d,structs:u,enums:m,abiJsonString:s,hexlifiedBinString:o,capitalizedName:i,imports:y}})}function Dt(n){let{abis:t,outputDir:e}=n,r=[],i=t.find(s=>s.commonTypesInUse.length>0);t.forEach(s=>{let{name:a}=s,u={path:`${e}/factories/${a}__factory.ts`,contents:Xt({abi:s})};r.push(u)});let o={path:`${e}/index.ts`,contents:v({abis:t})};if(r.push(o),i){let a={path:(0,Ut.join)(e,"common.d.ts"),contents:D()};r.push(a)}return r}var zt=require("path");var vt=`{{header}}

{{#if imports}}
import {
{{#each imports}}
  {{this}},
{{/each}}
} from 'fuels';
{{/if}}

{{#if commonTypesInUse}}
import type { {{commonTypesInUse}} } from "./common";
{{/if}}


{{#each enums}}
export type {{structName}}Input = Enum<{ {{inputValues}} }>;
{{#if recycleRef}}
export type {{structName}}Output = {{structName}}Input;
{{else}}
export type {{structName}}Output = Enum<{ {{outputValues}} }>;
{{/if}}
{{/each}}


{{#each structs}}
export type {{structName}}Input{{typeAnnotations}} = { {{inputValues}} };
{{#if recycleRef}}
export type {{structName}}Output{{typeAnnotations}} = {{structName}}Input{{typeAnnotations}};
{{else}}
export type {{structName}}Output{{typeAnnotations}} = { {{outputValues}} };
{{/if}}
{{/each}}

type {{capitalizedName}}Inputs = [{{inputs}}];
type {{capitalizedName}}Output = {{output}};

const _abi = {{abiJsonString}}

const _bin = '{{hexlifiedBinString}}'

export class {{capitalizedName}}__factory {

  static readonly abi = _abi
  static readonly bin = _bin

  static createInstance(wallet: Account) {

    const { abi, bin } = {{capitalizedName}}__factory

    const script = new Script<
      {{capitalizedName}}Inputs,
      {{capitalizedName}}Output
    >(bin, abi, wallet);

    return script;

  }

}
`;function Vt(n){let{abi:t}=n,{types:e}=t,{rawContents:r,name:i,hexlifiedBinContents:o}=n.abi,s=JSON.stringify(r,null,2),a=t.functions.find(f=>f.name==="main");if(!a)throw new Error("ABI doesn't have a 'main()' method.");let{enums:m}=V({types:e}),{structs:u}=B({types:e}),{imports:y}=z({types:e,baseMembers:["Script","Account"]}),{prefixedInputs:b,output:d}=a.attributes;return A({template:vt,data:{inputs:b,output:d,structs:u,enums:m,abiJsonString:s,hexlifiedBinString:o,capitalizedName:i,imports:y}})}function Bt(n){let{abis:t,outputDir:e}=n,r=[],i=t.find(s=>s.commonTypesInUse.length>0);t.forEach(s=>{let{name:a}=s,u={path:`${e}/factories/${a}__factory.ts`,contents:Vt({abi:s})};r.push(u)});let o={path:`${e}/index.ts`,contents:v({abis:t})};if(r.push(o),i){let a={path:(0,zt.join)(e,"common.d.ts"),contents:D()};r.push(a)}return r}var kt=J(require("lodash.upperfirst"));function nt(n){let{abiFilepath:t,binFilepath:e,binExists:r,programType:i}=n,o=i==="script";if(!r&&o)throw new Error([`Could not find BIN file for counterpart ${(0,kt.default)(i)} ABI.`,`  - ABI: ${t}`,`  - BIN: ${e}`,i].join(`
`))}var it=class{abis;abiFiles;binFiles;outputDir;files;constructor(t){let{abiFiles:e,binFiles:r,outputDir:i,programType:o}=t;this.outputDir=i,this.abiFiles=e,this.binFiles=r,this.abis=this.abiFiles.map(s=>{let a=s.path.replace("-abi.json",".bin"),m=this.binFiles.find(({path:y})=>y===a);return m||nt({abiFilepath:s.path,binExists:!!m,binFilepath:a,programType:o}),new et({filepath:s.path,rawContents:JSON.parse(s.contents),hexlifiedBinContents:m==null?void 0:m.contents,outputDir:i,programType:o})}),this.files=this.getAssembledFiles({programType:o})}getAssembledFiles(t){let{abis:e,outputDir:r}=this,{programType:i}=t;switch(i){case"contract":return Mt({abis:e,outputDir:r});case"script":return Bt({abis:e,outputDir:r});case"predicate":return Dt({abis:e,outputDir:r});default:throw new Error(`Invalid Typegen programType: ${i}`)}}};function Qt(n){let{cwd:t,inputs:e,output:r,silent:i,programType:o,filepaths:s}=n,a=(0,Kt.basename)(t),{log:m}=console;i&&(m=()=>({}));let u=[];if(!(s!=null&&s.length)&&(e==null?void 0:e.length))u=e.flatMap(f=>(0,Jt.sync)(f,{cwd:t}));else if(s!=null&&s.length)u=s;else throw new Error("You need to inform at least one parameter: `input` or `filepaths`");let y=u.map(f=>({path:f,contents:(0,_.readFileSync)(f,"utf-8")})),d=o==="script"?u.map(f=>{let x=f.replace("-abi.json",".bin"),re=(0,_.existsSync)(x);return nt({abiFilepath:f,binFilepath:x,binExists:re,programType:o}),{path:x,contents:(0,jt.hexlify)((0,_.readFileSync)(x))}}):[],l=new it({outputDir:r,abiFiles:y,binFiles:d,programType:o});m(`Generating files..
`),qt.default.sync(`${r}/factories`),l.files.forEach(f=>{Yt.default.sync(f.path),(0,_.writeFileSync)(f.path,f.contents);let x=new RegExp(`^.+${a}/`,"m");m(` - ${f.path.replace(x,"")}`)}),m(`
Done.\u26A1`)}function Zt(n){let{contract:t,script:e,predicate:r}=n;return t||!t&&!e&&!r?"contract":r?"predicate":"script"}function te(n){let{inputs:t,output:e,silent:r,contract:i,script:o,predicate:s}=n,a=process.cwd(),m=Zt({contract:i,script:o,predicate:s});Qt({cwd:a,inputs:t,output:e,programType:m,silent:!!r})}function ee(n){n.requiredOption("-i, --inputs <path|glob...>","input paths/globals to your abi json files").requiredOption("-o, --output <dir>","directory path for generated files").addOption(new k.Option("-c, --contract","generate types for Contracts [default]").conflicts(["script","predicate"]).implies({script:void 0,predicate:void 0})).addOption(new k.Option("-s, --script","generate types for Scripts").conflicts(["contract","predicate"]).implies({contract:void 0,predicate:void 0})).addOption(new k.Option("-p, --predicate","generate types for Predicates").conflicts(["contract","script"]).implies({contract:void 0,script:void 0})).option("-S, --silent","omit output messages").action(te)}function he(n){let t=new k.Command,{argv:e,programName:r}=n;t.name(r),t.version(Wt.versions.FUELS),t.usage("-i ../out/*-abi.json -o ./generated/"),ee(t),t.parse(e)}0&&(module.exports={configureCliOptions,resolveProgramType,run,runCliAction});
