// @fuel-ts/abi-typegen/dist/runTypegen.js
"use strict";var Qt=Object.create;var v=Object.defineProperty;var Wt=Object.getOwnPropertyDescriptor;var Zt=Object.getOwnPropertyNames;var te=Object.getPrototypeOf,ee=Object.prototype.hasOwnProperty;var re=(n,t,e)=>t in n?v(n,t,{enumerable:!0,configurable:!0,writable:!0,value:e}):n[t]=e;var ie=(n,t)=>{for(var e in t)v(n,e,{get:t[e],enumerable:!0})},Tt=(n,t,e,r)=>{if(t&&typeof t=="object"||typeof t=="function")for(let i of Zt(t))!ee.call(n,i)&&i!==e&&v(n,i,{get:()=>t[i],enumerable:!(r=Wt(t,i))||r.enumerable});return n};var j=(n,t,e)=>(e=n!=null?Qt(te(n)):{},Tt(t||!n||!n.__esModule?v(e,"default",{value:n,enumerable:!0}):e,n)),ne=n=>Tt(v({},"__esModule",{value:!0}),n);var p=(n,t,e)=>(re(n,typeof t!="symbol"?t+"":t,e),e);var fe={};ie(fe,{runTypegen:()=>le});module.exports=ne(fe);var vt=require("@ethersproject/bytes"),_=require("fs"),jt=require("glob"),Jt=j(require("mkdirp")),qt=require("path"),Kt=j(require("rimraf"));var It=j(require("lodash.upperfirst"));function dt(n){let e=[r=>r.replace(/\s+/g,"-"),r=>r.replace(/\./g,"-"),r=>r.replace(/_/g,"-"),r=>r.replace(/-[a-z]/g,i=>i.slice(-1).toUpperCase()),r=>r.replace(/-/g,""),r=>r.replace(/^\d+/,""),r=>(0,It.default)(r)].reduce((r,i)=>i(r),n);if(e==="")throw new Error(`Can't guess class name, please rename file: ${n}`);return e}function d(n){let{types:t,typeId:e}=n,r=t.find(({rawAbiType:{typeId:i}})=>i===e);if(!r)throw new Error(`Type ID not found: ${e}.`);return r.parseComponentsAttributes({types:t}),r}function h(n){let{types:t,typeArguments:e,parentTypeId:r,target:i}=n,o=`${i}Label`,s=[],a,m;r!==void 0&&(a=d({types:t,typeId:r}),m=a.attributes[o]),e.forEach(y=>{let b,I=y.type;try{b=d({types:t,typeId:I}).attributes[o]}catch{b="void"}if(y.typeArguments){let l=h({types:t,target:i,parentTypeId:y.type,typeArguments:y.typeArguments});s.push(l)}else s.push(`${b}`)});let u=s.join(", ");return m&&(u=`${m}<${u}>`),u}var K=class{name;types;rawAbiFunction;attributes;constructor(t){this.rawAbiFunction=t.rawAbiFunction,this.types=t.types,this.name=t.rawAbiFunction.name,this.attributes={inputs:this.bundleInputTypes(),output:this.bundleOutputTypes(),prefixedInputs:this.bundleInputTypes(!0)}}bundleInputTypes(t=!1){let{types:e}=this;return this.rawAbiFunction.inputs.map(i=>{let{name:o,type:s,typeArguments:a}=i,m=d({types:e,typeId:s}),u;return a?u=h({types:e,target:"input",parentTypeId:s,typeArguments:a}):u=m.attributes.inputLabel,t?`${o}: ${u}`:u}).join(", ")}bundleOutputTypes(){return h({types:this.types,target:"output",typeArguments:[this.rawAbiFunction.output]})}getDeclaration(){let{name:t}=this,{prefixedInputs:e,output:r}=this.attributes;return`${t}: InvokeFunction<[${e}], ${r}>`}};function ht(n){let{types:t,rawAbiFunction:e}=n;return new K({types:t,rawAbiFunction:e})}function gt(n){let{types:t,rawAbiFunctions:e}=n;return e.map(i=>ht({types:t,rawAbiFunction:i}))}var T=class{rawAbiType;attributes;requireImportFromFuels;constructor(t){this.rawAbiType=t.rawAbiType,this.attributes={inputLabel:"unknown",outputLabel:"unknown"},this.requireImportFromFuels=!1}};var Q=class extends T{name="array";static isSuitableFor(t){return Q.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){var u,y;let{types:e}=t,{type:r}=this.rawAbiType,i=Number((u=r.match(Q.MATCH_REGEX))==null?void 0:u[1]),o=[],s=[];(y=this.rawAbiType.components)==null||y.forEach(b=>{let{type:I,typeArguments:l}=b;if(l){let f=h({types:e,typeArguments:l,parentTypeId:I,target:"input"}),x=h({types:e,typeArguments:l,parentTypeId:I,target:"output"});o.push(f),s.push(x)}else{let{attributes:f}=d({types:e,typeId:I});o.push(f.inputLabel),s.push(f.outputLabel)}});let a=Array(i).fill(o[0]).join(", "),m=Array(i).fill(s[0]).join(", ");return this.attributes={inputLabel:`[${a}]`,outputLabel:`[${m}]`},this.attributes}},E=Q;p(E,"swayType","[_; 2]"),p(E,"MATCH_REGEX",/^\[_; ([0-9]+)\]$/m);var it=class extends T{name="str";static isSuitableFor(t){return it.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){return this.attributes={inputLabel:"string",outputLabel:"string"},this.attributes}},F=it;p(F,"swayType","str[3]"),p(F,"MATCH_REGEX",/^str\[(.+)\]$/m);var nt=class extends F{name="b256";static isSuitableFor(t){return nt.MATCH_REGEX.test(t.type)}},w=nt;p(w,"swayType","b256"),p(w,"MATCH_REGEX",/^b256$/m);var st=class extends w{name="b512";static isSuitableFor(t){return st.MATCH_REGEX.test(t.type)}},$=st;p($,"swayType","struct B512"),p($,"MATCH_REGEX",/^struct B512$/m);var ot=class extends T{name="bool";static isSuitableFor(t){return ot.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){return this.attributes={inputLabel:"boolean",outputLabel:"boolean"},this.attributes}},L=ot;p(L,"swayType","bool"),p(L,"MATCH_REGEX",/^bool$/m);function G(n){var i;let{rawAbiType:t,regex:e}=n,r=(i=t.type.match(n.regex))==null?void 0:i[1];if(!r){let o=`Couldn't extract struct name with: '${e}'.

`;throw o+=`Check your JSON ABI.

[source]
`,o+=`${JSON.stringify(t,null,2)}`,new Error(o)}return r}var J=class extends T{name="enum";static isSuitableFor(t){let e=J.MATCH_REGEX.test(t.type),r=J.IGNORE_REGEX.test(t.type);return e&&!r}parseComponentsAttributes(t){let e=this.getStructName();return this.attributes={structName:e,inputLabel:`${e}Input`,outputLabel:`${e}Output`},this.attributes}getStructName(){return G({rawAbiType:this.rawAbiType,regex:J.MATCH_REGEX})}getStructContents(t){let{types:e,target:r}=t,{components:i}=this.rawAbiType,o=i,s=`${r}Label`;return o.map(m=>{let{name:u,type:y}=m;if(y===0)return`${u}: []`;let{attributes:b}=d({types:e,typeId:y});return`${u}: ${b[s]}`}).join(", ")}},C=J;p(C,"swayType","enum MyEnumName"),p(C,"MATCH_REGEX",/^enum (.+)$/m),p(C,"IGNORE_REGEX",/^enum Option$/m);var W=class extends T{name="generic";static isSuitableFor(t){return W.MATCH_REGEX.test(t.type)}getStructName(){return G({rawAbiType:this.rawAbiType,regex:W.MATCH_REGEX})}parseComponentsAttributes(t){let e=this.getStructName();return this.attributes={inputLabel:e,outputLabel:e},this.attributes}},O=W;p(O,"swayType","generic T"),p(O,"MATCH_REGEX",/^generic ([^\s]+)$/m);var pt=class extends T{name="option";static isSuitableFor(t){return pt.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){return this.attributes={inputLabel:"Option",outputLabel:"Option"},this.attributes}},M=pt;p(M,"swayType","enum Option"),p(M,"MATCH_REGEX",/^enum Option$/m);var at=class extends T{name="u8";constructor(t){super(t),this.attributes={inputLabel:"BigNumberish",outputLabel:"number"},this.requireImportFromFuels=!0}static isSuitableFor(t){return at.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){return this.attributes}},g=at;p(g,"swayType","u8"),p(g,"MATCH_REGEX",/^u8$/m);var ct=class extends g{name="u64";parseComponentsAttributes(t){return this.attributes={inputLabel:"BigNumberish",outputLabel:"BN"},this.attributes}static isSuitableFor(t){return ct.MATCH_REGEX.test(t.type)}},R=ct;p(R,"swayType","u64"),p(R,"MATCH_REGEX",/^u64$/m);var mt=class extends R{name="rawUntypedPtr";static isSuitableFor(t){return mt.MATCH_REGEX.test(t.type)}},P=mt;p(P,"swayType","raw untyped ptr"),p(P,"MATCH_REGEX",/^raw untyped ptr$/m);var q=class extends T{name="struct";static isSuitableFor(t){let e=q.MATCH_REGEX.test(t.type),r=q.IGNORE_REGEX.test(t.type);return e&&!r}parseComponentsAttributes(t){let e=this.getStructName();return this.attributes={structName:e,inputLabel:`${e}Input`,outputLabel:`${e}Output`},this.attributes}getStructName(){return G({rawAbiType:this.rawAbiType,regex:q.MATCH_REGEX})}getStructContents(t){let{types:e,target:r}=t,{components:i}=this.rawAbiType;return i.map(a=>{let{name:m,type:u,typeArguments:y}=a,b=d({types:e,typeId:u}),I;if(y)I=h({types:e,target:r,parentTypeId:u,typeArguments:y});else{let l=`${r}Label`;I=b.attributes[l]}return`${m}: ${I}`}).join(", ")}getStructDeclaration(t){let{types:e}=t,{typeParameters:r}=this.rawAbiType;return r?`<${r.map(s=>d({types:e,typeId:s})).map(({attributes:{inputLabel:s}})=>s).join(", ")}>`:""}},N=q;p(N,"swayType","struct MyStruct"),p(N,"MATCH_REGEX",/^struct (.+)$/m),p(N,"IGNORE_REGEX",/^struct (Vec|RawVec)$/m);var ut=class extends T{name="tupple";static isSuitableFor(t){return ut.MATCH_REGEX.test(t.type)}parseComponentsAttributes(t){var o;let{types:e}=t,r=[],i=[];return(o=this.rawAbiType.components)==null||o.forEach(s=>{let{type:a,typeArguments:m}=s;if(m){let u=h({types:e,typeArguments:m,parentTypeId:a,target:"input"}),y=h({types:e,typeArguments:m,parentTypeId:a,target:"output"});r.push(u),i.push(y)}else{let{attributes:u}=d({types:e,typeId:a});r.push(u.inputLabel),i.push(u.outputLabel)}}),this.attributes={inputLabel:`[${r.join(", ")}]`,outputLabel:`[${i.join(", ")}]`},this.attributes}},H=ut;p(H,"swayType","(_, _, _)"),p(H,"MATCH_REGEX",/^\([_,\s]+\)$/m);var yt=class extends g{name="u16";static isSuitableFor(t){return yt.MATCH_REGEX.test(t.type)}},X=yt;p(X,"swayType","u16"),p(X,"MATCH_REGEX",/^u16$/m);var lt=class extends g{name="u32";static isSuitableFor(t){return lt.MATCH_REGEX.test(t.type)}},U=lt;p(U,"swayType","u32"),p(U,"MATCH_REGEX",/^u32$/m);var Z=class extends E{name="vector";static isSuitableFor(t){let e=Z.MATCH_REGEX.test(t.type),r=Z.IGNORE_REGEX.test(t.type);return e&&!r}parseComponentsAttributes(t){return this.attributes={inputLabel:"Vec",outputLabel:"Vec"},this.attributes}},S=Z;p(S,"swayType","struct Vec"),p(S,"MATCH_REGEX",/^struct Vec/m),p(S,"IGNORE_REGEX",/^struct RawVec$/m);var At=[E,w,$,L,C,O,M,P,F,N,H,X,U,R,g,S];function xt(n){let{rawAbiType:t}=n,{type:e}=t,r=At.find(i=>i.isSuitableFor({type:e}));if(!r)throw new Error(`Type not supported: ${e}`);return new r(n)}function Et(n){return["()","struct RawVec"].indexOf(n.type)>=0}function Ft(n){let t=[];return n.rawAbiTypes.forEach(e=>{let{type:r}=e;if(!Et({type:r})){let o=xt({rawAbiType:e});t.push(o)}}),t.forEach(e=>{e.parseComponentsAttributes({types:t})}),t}var tt=class{name;programType;filepath;outputDir;commonTypesInUse=[];rawContents;hexlifiedBinContents;types;functions;constructor(t){let{filepath:e,outputDir:r,rawContents:i,hexlifiedBinContents:o,programType:s}=t,a=/([^/]+)-abi\.json$/m,m=e.match(a);if(!m||m.length===0)throw new Error(`Could not parse name from abi file: ${e}`);let y=`${dt(m[1])}Abi`;this.name=y,this.programType=s,this.filepath=e,this.rawContents=i,this.hexlifiedBinContents=o,this.outputDir=r;let{types:b,functions:I}=this.parse();this.types=b,this.functions=I,this.computeCommonTypesInUse()}parse(){let{types:t,functions:e}=this.rawContents,r=Ft({rawAbiTypes:t}),i=gt({rawAbiFunctions:e,types:r});return{types:r,functions:i}}computeCommonTypesInUse(){let t={option:"Option",enum:"Enum",vector:"Vec"};this.commonTypesInUse=[],Object.keys(t).forEach(e=>{if(!!this.types.find(i=>i.name===e)){let i=t[e];this.commonTypesInUse.push(i)}})}};var Ot=require("path");var Rt=require("@fuel-ts/versions"),ft=require("handlebars");var wt=`/* Autogenerated file. Do not edit manually. */

/* tslint:disable */
/* eslint-disable */

/*
  Fuels version: {{FUELS}}
  Forc version: {{FORC}}
  Fuel-Core version: {{FUEL_CORE}}
*/
`;function A(n){let{data:t,template:e}=n,r={strict:!0,noEscape:!0},i=(0,ft.compile)(e,r),o=(0,ft.compile)(wt,r);return i({...t,header:o(Rt.versions)}).replace(/[\n]{3,}/gm,`

`)}var Ct=`{{header}}

/*
  Mimics Sway Enum, requires at least one Key-Value but
  does not raise error on multiple pairs.
  This is done in the abi-coder
*/
export type Enum<T, U = { [K in keyof T]: Pick<T, K> }> = Partial<T> &
  U[keyof U];

/*
  Mimics Sway Option and Vectors.
  Vectors are treated like arrays in Typescript.
*/
export type Option<T> = T | undefined;

export type Vec<T> = T[];
`;function D(){return A({template:Ct})}var Nt=`{{header}}

{{#if isGeneratingContracts}}
{{#each abis}}
export type { {{name}} } from './{{name}}';
{{/each}}
{{/if}}

{{#each abis}}
export { {{name}}__factory } from './factories/{{name}}__factory';
{{/each}}
`;function V(n){let{abis:t}=n,e=t[0].programType==="contract";return A({template:Nt,data:{abis:t,isGeneratingContracts:e}})}function z(n){let{types:t}=n;return{enums:t.filter(r=>r.name==="enum").map(r=>{let i=r,o=i.getStructName(),s=i.getStructContents({types:t,target:"input"}),a=i.getStructContents({types:t,target:"output"});return{structName:o,inputValues:s,outputValues:a,recycleRef:s===a}})}}var St=j(require("lodash.uniq")),ae=(n,t)=>n.toLowerCase().localeCompare(t.toLowerCase());function B(n){let{types:t,baseMembers:e=[]}=n,i=t.filter(s=>s.requireImportFromFuels).map(s=>s.attributes.inputLabel),o=(0,St.default)(e.concat(i).sort(ae));return{imports:o.length?o:void 0}}function k(n){let{types:t}=n;return{structs:t.filter(r=>r.name==="struct").map(r=>{let i=r,o=i.getStructName(),s=i.getStructContents({types:t,target:"input"}),a=i.getStructContents({types:t,target:"output"}),m=i.getStructDeclaration({types:t});return{structName:o,typeAnnotations:m,inputValues:s,outputValues:a,recycleRef:s===a}})}}var _t=`{{header}}

{{#if imports}}
import type {
{{#each imports}}
  {{this}},
{{/each}}
} from 'fuels';
{{/if}}

{{#if commonTypesInUse}}
import type { {{commonTypesInUse}} } from "./common";
{{/if}}

{{#each enums}}
export type {{structName}}Input = Enum<{ {{inputValues}} }>;
{{#if recycleRef}}
export type {{structName}}Output = {{structName}}Input;
{{else}}
export type {{structName}}Output = Enum<{ {{outputValues}} }>;
{{/if}}
{{/each}}


{{#each structs}}
export type {{structName}}Input{{typeAnnotations}} = { {{inputValues}} };
{{#if recycleRef}}
export type {{structName}}Output{{typeAnnotations}} = {{structName}}Input{{typeAnnotations}};
{{else}}
export type {{structName}}Output{{typeAnnotations}} = { {{outputValues}} };
{{/if}}
{{/each}}


interface {{capitalizedName}}Interface extends Interface {
  functions: {
    {{#each functionsFragments}}
    {{this}}: FunctionFragment;
    {{/each}}
  };

  {{#each encoders}}
  encodeFunctionData(functionFragment: '{{functionName}}', values: [{{input}}]): Uint8Array;
  {{/each}}

  {{#each decoders}}
  decodeFunctionData(functionFragment: '{{functionName}}', data: BytesLike): DecodedValue;
  {{/each}}
}


export class {{capitalizedName}} extends Contract {
  interface: {{capitalizedName}}Interface;
  functions: {
    {{#each functionsTypedefs}}
    {{this}};
    {{/each}}
  };
}
`;function $t(n){let{name:t,types:e,functions:r,commonTypesInUse:i}=n.abi,o=r.map(l=>l.getDeclaration()),s=r.map(l=>l.name),a=r.map(l=>({functionName:l.name,input:l.attributes.inputs})),m=r.map(l=>({functionName:l.name})),{enums:u}=z({types:e}),{structs:y}=k({types:e}),{imports:b}=B({types:e,baseMembers:["Interface","FunctionFragment","DecodedValue","Contract","BytesLike","InvokeFunction","BN"]});return A({template:_t,data:{capitalizedName:t,commonTypesInUse:i.join(", "),functionsTypedefs:o,functionsFragments:s,encoders:a,decoders:m,structs:y,enums:u,imports:b}})}var Lt=`{{header}}

import { Interface, Contract } from "fuels";
import type { Provider, Account, AbstractAddress } from "fuels";
import type { {{capitalizedName}}, {{capitalizedName}}Interface } from "../{{capitalizedName}}";

const _abi = {{abiJsonString}}

export class {{capitalizedName}}__factory {
  static readonly abi = _abi
  static createInterface(): {{capitalizedName}}Interface {
    return new Interface(_abi) as unknown as {{capitalizedName}}Interface
  }
  static connect(
    id: string | AbstractAddress,
    accountOrProvider: Account | Provider
  ): {{capitalizedName}} {
    return new Contract(id, _abi, accountOrProvider) as unknown as {{capitalizedName}}
  }
}
`;function Gt(n){let{name:t,rawContents:e}=n.abi,r=JSON.stringify(e,null,2);return A({template:Lt,data:{capitalizedName:t,abiJsonString:r}})}function Mt(n){let{abis:t,outputDir:e}=n,r=[],i=t.find(s=>s.commonTypesInUse.length>0);t.forEach(s=>{let{name:a}=s,m=`${e}/${a}.d.ts`,u=`${e}/factories/${a}__factory.ts`,y={path:m,contents:$t({abi:s})},b={path:u,contents:Gt({abi:s})};r.push(y),r.push(b)});let o={path:`${e}/index.ts`,contents:V({abis:t})};if(r.push(o),i){let a={path:(0,Ot.join)(e,"common.d.ts"),contents:D()};r.push(a)}return r}var Xt=require("path");var Pt=`{{header}}

{{#if imports}}
import {
{{#each imports}}
  {{this}},
{{/each}}
} from 'fuels';
{{/if}}

{{#if commonTypesInUse}}
import type { {{commonTypesInUse}} } from "./common";
{{/if}}


{{#each enums}}
export type {{structName}}Input = Enum<{ {{inputValues}} }>;
{{#if recycleRef}}
export type {{structName}}Output = {{structName}}Input;
{{else}}
export type {{structName}}Output = Enum<{ {{outputValues}} }>;
{{/if}}
{{/each}}


{{#each structs}}
export type {{structName}}Input{{typeAnnotations}} = { {{inputValues}} };
{{#if recycleRef}}
export type {{structName}}Output{{typeAnnotations}} = {{structName}}Input{{typeAnnotations}};
{{else}}
export type {{structName}}Output{{typeAnnotations}} = { {{outputValues}} };
{{/if}}
{{/each}}

type {{capitalizedName}}Inputs = [{{inputs}}];

const _abi = {{abiJsonString}}

const _bin = '{{hexlifiedBinString}}'

export class {{capitalizedName}}__factory {

  static readonly abi = _abi
  static readonly bin = _bin;

  static createInstance(provider?: Provider) {

    const { abi, bin } = {{capitalizedName}}__factory

    const predicate = new Predicate(bin, abi, provider);

    return predicate;

  }

}
`;function Ht(n){let{abi:t}=n,{types:e}=t,{rawContents:r,name:i,hexlifiedBinContents:o}=n.abi,s=JSON.stringify(r,null,2),a=t.functions.find(f=>f.name==="main");if(!a)throw new Error("ABI doesn't have a 'main()' method.");let{enums:m}=z({types:e}),{structs:u}=k({types:e}),{imports:y}=B({types:e,baseMembers:["Predicate","Provider"]}),{prefixedInputs:b,output:I}=a.attributes;return A({template:Pt,data:{inputs:b,output:I,structs:u,enums:m,abiJsonString:s,hexlifiedBinString:o,capitalizedName:i,imports:y}})}function Ut(n){let{abis:t,outputDir:e}=n,r=[],i=t.find(s=>s.commonTypesInUse.length>0);t.forEach(s=>{let{name:a}=s,u={path:`${e}/factories/${a}__factory.ts`,contents:Ht({abi:s})};r.push(u)});let o={path:`${e}/index.ts`,contents:V({abis:t})};if(r.push(o),i){let a={path:(0,Xt.join)(e,"common.d.ts"),contents:D()};r.push(a)}return r}var zt=require("path");var Dt=`{{header}}

{{#if imports}}
import {
{{#each imports}}
  {{this}},
{{/each}}
} from 'fuels';
{{/if}}

{{#if commonTypesInUse}}
import type { {{commonTypesInUse}} } from "./common";
{{/if}}


{{#each enums}}
export type {{structName}}Input = Enum<{ {{inputValues}} }>;
{{#if recycleRef}}
export type {{structName}}Output = {{structName}}Input;
{{else}}
export type {{structName}}Output = Enum<{ {{outputValues}} }>;
{{/if}}
{{/each}}


{{#each structs}}
export type {{structName}}Input{{typeAnnotations}} = { {{inputValues}} };
{{#if recycleRef}}
export type {{structName}}Output{{typeAnnotations}} = {{structName}}Input{{typeAnnotations}};
{{else}}
export type {{structName}}Output{{typeAnnotations}} = { {{outputValues}} };
{{/if}}
{{/each}}

type {{capitalizedName}}Inputs = [{{inputs}}];
type {{capitalizedName}}Output = {{output}};

const _abi = {{abiJsonString}}

const _bin = '{{hexlifiedBinString}}'

export class {{capitalizedName}}__factory {

  static readonly abi = _abi
  static readonly bin = _bin

  static createInstance(wallet: Account) {

    const { abi, bin } = {{capitalizedName}}__factory

    const script = new Script<
      {{capitalizedName}}Inputs,
      {{capitalizedName}}Output
    >(bin, abi, wallet);

    return script;

  }

}
`;function Vt(n){let{abi:t}=n,{types:e}=t,{rawContents:r,name:i,hexlifiedBinContents:o}=n.abi,s=JSON.stringify(r,null,2),a=t.functions.find(f=>f.name==="main");if(!a)throw new Error("ABI doesn't have a 'main()' method.");let{enums:m}=z({types:e}),{structs:u}=k({types:e}),{imports:y}=B({types:e,baseMembers:["Script","Account"]}),{prefixedInputs:b,output:I}=a.attributes;return A({template:Dt,data:{inputs:b,output:I,structs:u,enums:m,abiJsonString:s,hexlifiedBinString:o,capitalizedName:i,imports:y}})}function Bt(n){let{abis:t,outputDir:e}=n,r=[],i=t.find(s=>s.commonTypesInUse.length>0);t.forEach(s=>{let{name:a}=s,u={path:`${e}/factories/${a}__factory.ts`,contents:Vt({abi:s})};r.push(u)});let o={path:`${e}/index.ts`,contents:V({abis:t})};if(r.push(o),i){let a={path:(0,zt.join)(e,"common.d.ts"),contents:D()};r.push(a)}return r}var kt=j(require("lodash.upperfirst"));function et(n){let{abiFilepath:t,binFilepath:e,binExists:r,programType:i}=n,o=i==="script";if(!r&&o)throw new Error([`Could not find BIN file for counterpart ${(0,kt.default)(i)} ABI.`,`  - ABI: ${t}`,`  - BIN: ${e}`,i].join(`
`))}var rt=class{abis;abiFiles;binFiles;outputDir;files;constructor(t){let{abiFiles:e,binFiles:r,outputDir:i,programType:o}=t;this.outputDir=i,this.abiFiles=e,this.binFiles=r,this.abis=this.abiFiles.map(s=>{let a=s.path.replace("-abi.json",".bin"),m=this.binFiles.find(({path:y})=>y===a);return m||et({abiFilepath:s.path,binExists:!!m,binFilepath:a,programType:o}),new tt({filepath:s.path,rawContents:JSON.parse(s.contents),hexlifiedBinContents:m==null?void 0:m.contents,outputDir:i,programType:o})}),this.files=this.getAssembledFiles({programType:o})}getAssembledFiles(t){let{abis:e,outputDir:r}=this,{programType:i}=t;switch(i){case"contract":return Mt({abis:e,outputDir:r});case"script":return Bt({abis:e,outputDir:r});case"predicate":return Ut({abis:e,outputDir:r});default:throw new Error(`Invalid Typegen programType: ${i}`)}}};function le(n){let{cwd:t,inputs:e,output:r,silent:i,programType:o,filepaths:s}=n,a=(0,qt.basename)(t),{log:m}=console;i&&(m=()=>({}));let u=[];if(!(s!=null&&s.length)&&(e==null?void 0:e.length))u=e.flatMap(f=>(0,jt.sync)(f,{cwd:t}));else if(s!=null&&s.length)u=s;else throw new Error("You need to inform at least one parameter: `input` or `filepaths`");let y=u.map(f=>({path:f,contents:(0,_.readFileSync)(f,"utf-8")})),I=o==="script"?u.map(f=>{let x=f.replace("-abi.json",".bin"),Yt=(0,_.existsSync)(x);return et({abiFilepath:f,binFilepath:x,binExists:Yt,programType:o}),{path:x,contents:(0,vt.hexlify)((0,_.readFileSync)(x))}}):[],l=new rt({outputDir:r,abiFiles:y,binFiles:I,programType:o});m(`Generating files..
`),Jt.default.sync(`${r}/factories`),l.files.forEach(f=>{Kt.default.sync(f.path),(0,_.writeFileSync)(f.path,f.contents);let x=new RegExp(`^.+${a}/`,"m");m(` - ${f.path.replace(x,"")}`)}),m(`
Done.\u26A1`)}0&&(module.exports={runTypegen});
